tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/ScalaJSClassEmitter.scala

```diff
@@ -590,7 +679,7 @@ private[scalajs] final class ScalaJSClassEmitter(
         js.Function(List(objParam, depthParam), className match {
           case Definitions.ObjectClass =>
             val dataVarDef = genLet(Ident("data"), mutable = false, {
-              obj && (obj DOT "$classData")
+              obj && js.BracketSelect(envField("ClassData"), obj DOT "$typeTag")
```

Comment:
There's no need to go through `ClassData`. You have all the necessary information in `obj DOT "$typeTag"`. Indeed, all you need is to compute `arrayDepth` and `arrayBaseTag`.

You can replace the currently used expression `data.arrayBase["isPrimitive"]` by `arrayBaseTag < 9`.